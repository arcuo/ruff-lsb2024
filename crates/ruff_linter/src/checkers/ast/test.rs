use std::path::Path;

use ruff_python_index::Indexer;
use ruff_python_parser::{lexer::LexResult, Mode};
use ruff_python_semantic::Module;
use ruff_source_file::Locator;

use crate::{
    directives, linter::TokenSource, packaging::detect_package_root, settings::{flags::Noqa, LinterSettings}
};

use super::Checker;

// TODO: Should we use this?

/// Create a new [`Checker`] for the given source code
pub(crate) fn test_checker_state_after_snippet_traversal(source: &str) -> &Checker {
    let path = Path::new("");
    let locator = Locator::new(source);
    let tokens = TokenSource::Tokens(ruff_python_parser::tokenize_all(source, Mode::Module));
    let indexer = Indexer::from_tokens(&tokens, &locator);
    let settings = LinterSettings::default();
    let directives = directives::extract_directives(
        &tokens,
        directives::Flags::from_settings(&settings),
        &locator,
        &indexer,
    );
    let checker = Checker::new(
        &settings,
        &directives.noqa_line_for,
        Noqa::Disabled,
        &path,
        path.parent()
            .and_then(|parent| detect_package_root(parent, &settings.namespace_packages)),
        Module {
            kind: ruff_python_semantic::ModuleKind::Module,
            source: ruff_python_semantic::ModuleSource::Path(&[path.to_str().unwrap().to_string()]),
            python_ast: tokens.,
            name: path.to_string(),
        },
    );
    &checker
}
