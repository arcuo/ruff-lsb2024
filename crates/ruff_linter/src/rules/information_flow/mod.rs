//! Rules for information flow.
pub(crate) mod rules;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::test::test_path;
    use crate::{assert_messages, settings};

    // Basic rules
    #[test_case(Rule::IFMustIncludeVariableLabel, Path::new("IF001.py"))]
    #[test_case(Rule::IFMissingPrincipal, Path::new("IF002.py"))]
    // Explicit
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101.py"))] // TODO: split into multiple tests
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101_for.py"))] // TODO: split into multiple tests

    // Implicit
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_if.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_while.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_for.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_nested.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_assert.py"))]

    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("information_flow").join(path).as_path(),
            &settings::LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
