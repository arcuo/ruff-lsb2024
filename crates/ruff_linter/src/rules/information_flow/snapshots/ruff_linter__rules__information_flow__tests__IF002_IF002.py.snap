---
source: crates/ruff_linter/src/rules/information_flow/mod.rs
---
IF002.py:6:18: IF002 [*] Missing principal `charlie` in the principals list for `{ charlie }`
  |
4 | ## BAD
5 | 
6 | b = 0 # iflabel {charlie}
  |                  ^^^^^^^ IF002
7 | b2: int = 0 # iflabel {alice, charlie}
  |
  = help: Add principal `charlie` to the top of the file `# ifprincipals {alice, bob, charlie}`

ℹ Unsafe fix
1   |-# ifprincipals {alice, bob}
  1 |+# ifprincipals { alice, bob, charlie }
2 2 | 
3 3 | 
4 4 | ## BAD

IF002.py:7:31: IF002 [*] Missing principal `charlie` in the principals list for `{ alice, charlie }`
  |
6 | b = 0 # iflabel {charlie}
7 | b2: int = 0 # iflabel {alice, charlie}
  |                               ^^^^^^^ IF002
8 | 
9 | ## GOOD
  |
  = help: Add principal `charlie` to the top of the file `# ifprincipals {alice, bob, charlie}`

ℹ Unsafe fix
1   |-# ifprincipals {alice, bob}
  1 |+# ifprincipals { alice, bob, charlie }
2 2 | 
3 3 | 
4 4 | ## BAD
